tests/test_connection.py

```diff
@@ -0,0 +1,189 @@
+
+    @mock.patch('requests.Session.post')
+    def test_make_request(self, post_mock):
+        response_mock = mock.MagicMock(
+            status_code=200,
+            headers={
+                'content-type': 'application/json'
+            },
+        )
+        response_mock.json.return_value = {'ok': 'ok'}
+        post_mock.return_value = response_mock
+
+        out = self.connection.make_request('POST', 'test')
+        self.assertTrue(post_mock.called)
+        self.assertEqual(out, response_mock.json.return_value)
+
+    def test_invalid_method_name(self):
+        with self.assertRaises(SyncanoValueError):
+            self.connection.make_request('INVALID', 'test')
+
+    @mock.patch('requests.Session.post')
+    def test_request_error(self, post_mock):
+        post_mock.return_value = mock.MagicMock(status_code=404, text='Invalid request')
+        self.assertFalse(post_mock.called)
+
+        with self.assertRaises(SyncanoRequestError):
+            self.connection.make_request('POST', 'test')
+
+        self.assertTrue(post_mock.called)
+
+    def test_is_authenticated(self):
+        self.assertFalse(self.connection.is_authenticated())
+        self.connection.api_key = 'xxxx'
+        self.assertTrue(self.connection.is_authenticated())
+
+    @mock.patch('syncano.connection.Connection.make_request')
+    def test_already_authenticated(self, make_request_mock):
+        self.assertIsNone(self.connection.api_key)
+        self.assertFalse(make_request_mock.called)
+
+        self.connection.api_key = 'Ala has cat'
+        out = self.connection.authenticate()
+
+        self.assertEqual(out, 'Ala has cat')
+        self.assertFalse(make_request_mock.called)
+
+    @mock.patch('syncano.connection.Connection.make_request')
+    def test_authentication_empty_email(self, make_request_mock):
+        self.assertFalse(make_request_mock.called)
+        with self.assertRaises(SyncanoValueError):
+            self.connection.authenticate()
+        self.assertFalse(make_request_mock.called)
+
+    @mock.patch('syncano.connection.Connection.make_request')
+    def test_authentication_empty_password(self, make_request_mock):
+        self.assertFalse(make_request_mock.called)
+        with self.assertRaises(SyncanoValueError):
+            self.connection.authenticate(email='dummy')
+        self.assertFalse(make_request_mock.called)
+
+    @mock.patch('requests.Session.post')
+    def test_invalid_credentials(self, post_mock):
+        post_mock.return_value = mock.MagicMock(status_code=401, text='Invalid email or password.')
+        self.assertFalse(post_mock.called)
+        self.assertIsNone(self.connection.api_key)
+
+        with self.assertRaises(SyncanoRequestError) as cm:
+            self.connection.authenticate(email='dummy', password='dummy')
+
+        self.assertEqual(cm.exception.status_code, 401)
+        self.assertEqual(cm.exception.message, 'Invalid email or password.')
```

Comment:
I really don't like pouring variables out of their scope, so I think that it would be better to test variable cm within 'with' block.